reconcile/external_resources/state.py

```diff
@@ -0,0 +1,246 @@
+    DELETED: str = "DELETED"
+    ABANDONED: str = "ABANDONED"
+    NOT_EXISTS: str = "NOT_EXISTS"
+    IN_PROGRESS: str = "IN_PROGRESS"
+    DELETE_IN_PROGRESS: str = "DELETE_IN_PROGRESS"
+    ERROR: str = "ERROR"
+
+
+class ExternalResourceState(BaseModel):
+    key: ExternalResourceKey
+    ts: datetime
+    resource_status: ResourceStatus
+    resource_digest: str = ""
+    reconciliation: Reconciliation
+    reconciliation_errors: int = 0
+
+
+class EnhancedJsonEncoder(JSONEncoder):
+    def default(self, obj: Any) -> Any:
+        if isinstance(obj, datetime):
+            return obj.isoformat()
+        elif dataclasses.is_dataclass(obj):
+            return dataclasses.asdict(obj)
+
+
+class ExternalResourcesStateManager:
+    def __init__(self, state: State, index_file_key: str):
+        self.state = state
+        self.index_file_key = index_file_key
+
+        try:
+            data = self.state[self.index_file_key]
+            states_list = parse_obj_as(list[ExternalResourceState], data)
+            self.index: dict[ExternalResourceKey, ExternalResourceState] = {
+                item.key: item for item in states_list
+            }
+        except Exception:
+            logging.info("No state file, creating a new one.")
+            self.index = {}
+
+    def _write_index_file(
+        self,
+    ) -> None:
+        data = [item.dict() for item in self.index.values()]
+        self.state[self.index_file_key] = data
+
+    def get_external_resource_state(
+        self, key: ExternalResourceKey
+    ) -> ExternalResourceState:
+        obj = self.index.get(
+            key,
+            ExternalResourceState(
+                key=key,
+                ts=datetime.now(timezone.utc),
+                resource_status=ResourceStatus.NOT_EXISTS,
+                reconciliation=Reconciliation(key=key),
+            ),
+        )
+        return obj
+
+    def set_external_resource_state(
+        self,
+        key: ExternalResourceKey,
+        state: ExternalResourceState,
+    ) -> None:
+        self.index[key] = state
+
+    def del_external_resource_state(self, key: ExternalResourceKey) -> None:
+        del self.index[key]
+
+    def get_all_resource_keys(self) -> list[ExternalResourceKey]:
+        return list[ExternalResourceKey](self.index.keys())
+
+    def save_state(self) -> None:
+        self._write_index_file()
+        # self._write_external_resource_states()
+
+
+class DynamoDBStateAdapater:
+    KEY_PROVISION_PROVIDER = "key.provision_provider"
+    KEY_PROVISIONER_NAME = "key.provisioner_name"
+    KEY_PROVIDER = "key.provider"
+    KEY_IDENTIFIER = "key.identifier"
+    RECONCILIATION_RESOURCE_DIGEST = "reconcilitation.resource_digest"
+    RECONCILIATION_IMAGE = "reconciliation.image"
+    RECONCILIATION_INPUT = "reconciliation.input"
+    RECONCILIATION_ACTION = "reconciliation.action"
+    RESOURCE_KEY = "resource_key"
+    RESOURCE_STATUS = "resource_status"
+    RESOURCE_DIGEST = "resource_digest"
```

Comment:
nit: In other places we do not use _digest_ but _hash_ for similar concepts. Perhaps stick to it?